### PR TITLE
Support self:// for Solaris ##io

### PR DESCRIPTION
**Pull request by devnexen**
enabled only on 64 bits unsupported on 32 bits with large files.
missing cfmakeraw function on this platform for socket et cons.

 

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

### CODE REVIEW

File: libr/io/p/io_self.c [hunk 1]
```diff
@@ -144,6 +148,52 @@ static int update_self_regions(RIO *io, int pid) {
...
+	char path[PATH_MAX];
+	ssize_t rd;
+	size_t hint;
+	int fd;
+
+	snprintf (path, sizeof (path), "/proc/%d/map", getpid());
+	hint = (1 << 20);
+	fd = open (path, O_RDONLY);
+	if (fd != -1) {
+		map = malloc (hint);
```
Comment (trufae) [inline libr/io/p/io_self.c:163]: define and assign in the same line if possible, same for fd and other variables

File: libr/io/p/io_self.c [hunk 2]
```diff
@@ -144,6 +148,52 @@ static int update_self_regions(RIO *io, int pid) {
...
 	}
+	return true;
+#elif __sun && defined _LP64
+	prmap_t *c, *map;
+	char path[PATH_MAX];
+	ssize_t rd;
+	size_t hint;
+	int fd;
+
+	snprintf (path, sizeof (path), "/proc/%d/map", getpid());
```
Comment (trufae) [inline libr/io/p/io_self.c:159]: missing space before ( in getpid

File: libr/include/r_cons.h [hunk 3]
```diff
@@ -1216,8 +1216,18 @@ typedef struct r_panels_root_t {
 	RPanelsRootState root_state;
 } RPanelsRoot;
 
-#ifdef __cplusplus
+
+#ifdef __sun
+static inline void cfmakeraw(struct termios *tm) {
```
Comment (trufae) [inline libr/include/r_cons.h:1221]: this function brings me a lot of memories x))

File: libr/io/p/io_self.c [hunk 4]
```diff
@@ -144,6 +148,52 @@ static int update_self_regions(RIO *io, int pid) {
...
+	hint = (1 << 20);
```
Comment (trufae) [inline libr/io/p/io_self.c:160]: check if hint is  less than 1

File: libr/io/p/io_self.c [hunk 5]
```diff
@@ -144,6 +148,52 @@ static int update_self_regions(RIO *io, int pid) {
...
+	fd = open (path, O_RDONLY);
+	if (fd != -1) {
+		map = malloc (hint);
+
+		while ((rd = pread(fd, map, hint, 0)) == hint) {
+			hint <<= 1;
+			map = realloc (map, hint);
+		}
+
+		for (c = map; rd; c ++, rd -= sizeof (prmap_t)) {
```
Comment (trufae) [inline libr/io/p/io_self.c:170]: rd can become negative adn we dont want that, use rd > 0 instead of rd

File: libr/io/p/io_self.c [hunk 6]
```diff
@@ -144,6 +148,52 @@ static int update_self_regions(RIO *io, int pid) {
...
+	size_t hint;
+	int fd;
+
+	snprintf (path, sizeof (path), "/proc/%d/map", getpid());
+	hint = (1 << 20);
+	fd = open (path, O_RDONLY);
+	if (fd != -1) {
+		map = malloc (hint);
+
+		while ((rd = pread(fd, map, hint, 0)) == hint) {
```
Comment (trufae) [inline libr/io/p/io_self.c:165]: missing spaace

File: libr/socket/run.c [hunk 7]
```diff
@@ -9,6 +9,7 @@
 #include <r_socket.h>
 #include <r_util.h>
 #include <r_lib.h>
+#include <r_cons.h>
```
Comment (trufae) [inline libr/socket/run.c:12]: why

File: libr/include/r_cons.h (hunk 3 above)
Comment (trufae) [inline libr/include/r_cons.h:1221]: but.. its not referenced from wanywhere :?

Comment (devnexen) [reply libr/include/r_cons.h:1221]: cons and run uses it tought, I got linkage issue this is how I spotted it.

File: libr/io/p/io_self.c [hunk 8]
```diff
@@ -144,6 +148,49 @@ static int update_self_regions(RIO *io, int pid) {
 		self_sections[self_sections_count].perm = 0;
 		self_sections_count++;
 	}
+	return true;
+#elif __sun && defined _LP64
+	char path[PATH_MAX];
+	snprintf (path, sizeof (path), "/proc/%d/map", getpid ());
+	size_t hint = (1 << 20);
+	int fd = open (path, O_RDONLY);
```
Comment (trufae) [inline libr/io/p/io_self.c:156]: if (fd == -1) return, to avoid nesting below

**Comment by trufae** [comment]
Im curious why you add such space in the checkboxes when opening PRs. is not the first time i see this :?

Screenshot 2020-07-02 at 15 24 13

**Comment by devnexen** [comment]
Ah sorry I did not realize

**Pull request merged**